### PR TITLE
docs: Add guidance for writing healthchecks

### LINKED ISSUE #708: Health check specification
It's become clear to me that we need a specification for health checks; they are consistently a point of discussion, confusion, and blockers for features that implement dynamic values. For example, #620, #621, #622, #683.

## Specification Requirements

Here's what I would like to see in the spec:

1. Start with the use case. Why would someone want to use a health check?
2. Based on that, define what a health check is.
3. Define it's responsibilities.
4. Define behavior boundaries (ex: delete behavior is _never_ an option). 

## Discussed Strategies

Here are some of the strategies we've discussed internally:

1. A health check should be light weight, it should _only_ verify that the service is reachable. It should not be concerned with configuration or permissions.
2. A health check should _never_ write test data that could pollute data in the downstream service. It should do everything possible to verify that errors will not be encountered when Vector is started with real data. This means requiring additional permissions and more. If the user does not want to opt into these permissions then they are effectively opting out of the health check, at which point they can disable the health check (#621)
3. A health check should send test data into the sink and mimic behavior as if it was started. It should _never_ deviate in behavior or require permissions outside those required to run Vector.

## Edge Cases

1. How do we handle configured values that include variables (`{{ field_name }}`). Do we randomly generate value? Do we require the user to provide health check values?
2. How do we handle sinks with strict schemas? This is only relevant if we go with option 3 above. But inserting random test data would always fail because it would not conform to the strict schema.

## Prior Art

TODO: fill in with strategies used by other tools.

## TODO

I think we can add this as it's own section in the [sinks page](https://github.com/timberio/vector/blob/master/docs/usage/configuration/sinks/README.md)

### PR DESCRIPTION
**Pull request by lukesteensen**
Closes #708 

Because the available options and user expectations can differ so much between sinks, this stops short of laying out a single strategy that should always be followed. Instead, it goes over the goals of healthchecks and lays out some of the common pitfalls we want to avoid. Finally, there's a checklist that should help guide new implementations.

### CODE REVIEW

File: docs/usage/configuration/sinks/README.md [hunk 1]
```diff
@@ -10,7 +10,7 @@ description: Receive and pull log and metric events into Vector
      scripts/generate/templates/docs/usage/configuration/sinks/README.md.erb
 -->
 
-# Sources
+# Sinks
```
Comment (binarylogic) [inline docs/usage/configuration/sinks/README.md]: Thanks for catching this

File: scripts/generate/templates/docs/usage/configuration/sinks/README.md.erb [hunk 2]
```diff
@@ -15,3 +15,52 @@ Sinks are last in the [pipeline][docs.pipelines], responsible for sending
...
+
+Sinks may implement a healthcheck as a means for validating their configuration
+against the envionment and external systems. Ideally, this allows the system to
+inform users of problems such as insufficient credentials, unreachable
+endpoints, non-existant tables, etc. They're not perfect, however, since it's
+impossible to exhaustively check for issues that may happen at runtime.
+
+### Guidelines for writing healthchecks
+
+When implementing healthchecks, we prefer false positives to false negatives.
```
Comment (binarylogic) [inline scripts/generate/templates/docs/usage/configuration/sinks/README.md.erb]: 👍 this is a good way to phrase it

File: scripts/generate/templates/docs/usage/configuration/sinks/README.md.erb [hunk 3]
```diff
@@ -15,3 +15,52 @@ Sinks are last in the [pipeline][docs.pipelines], responsible for sending
...
+doesn't need the ability to list all buckets, and a user that knows that may not
+have given it permission to do so. In that case, the healthcheck will fail due
+to bad credentials even through its credentials are sufficient for normal
+operation.
+
+This leads to a general strategy of mimicking what the sink itself does.
+Unfortunately, the fact that healthchecks don't have real events available to
+them leads to some limitations here. The most obvious example of this is with
+sinks where the exact target of a write depends on the value of some field in
+the event (e.g. an interpolated Kinesis stream name). It also pops up for sinks
```
Comment (binarylogic) [inline scripts/generate/templates/docs/usage/configuration/sinks/README.md.erb]: What would you propose for these circumstances then? Disable the health check?

Comment (lukesteensen) [reply scripts/generate/templates/docs/usage/configuration/sinks/README.md.erb]: Yeah, if we can't know where data is actually going to be sent I'm not sure what we can do that isn't a guess. I'll add some clarification that in cases like this it's ok to fall back to doing nothing.

File: scripts/generate/templates/docs/usage/configuration/sinks/README.md.erb [hunk 4]
```diff
@@ -15,3 +15,59 @@ Sinks are last in the [pipeline][docs.pipelines], responsible for sending
...
+cases, random test data is reasonably likely to trigger a potentially false
+negative result. Even in simpler cases, we need to think about the effects of
+writing test data and whether the user would find that surprising or invasive.
+The answer usually depends on the system we're interfacing with.
+
+In some cases, like the Kinesis example above, the right thing to do might be
+nothing at all. If we require dynamic information to figure out what entity
+(i.e. Kinesis stream in this case) that we're even dealing with, odds are very
+low that we'll be able to come up with a way to meaningfully validate that it's
+in working order. It's perfectly valid to have a healthcheck that falls back to
```
Comment (LucioFranco) [inline scripts/generate/templates/docs/usage/configuration/sinks/README.md.erb]: So I'm assuming this is kinda the main point, we are happy to do nothing at all if we can't reliably not cause false positives and data pollution? What worries me here is as we create more advanced sinks health checks will start to become almost irrelevant since there is a lot of dynamicness going on?

Comment (lukesteensen) [reply scripts/generate/templates/docs/usage/configuration/sinks/README.md.erb]: That's possible, yeah. I think there will always be popular sinks for which they're useful, but if turn out not to be a big deal that's ok too. Unless we want to compromise on the false positive and data pollution values, I don't think we really have another option.
